src/Workspaces/Remote/Core/Services/AssetService.cs

```diff
@@ -0,0 +1,165 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Execution;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// This service provide a way to get roslyn objects from checksum
+    /// 
+    /// TODO: change this service to workspace service
+    /// </summary>
+    internal class AssetService
+    {
+        private const int CleanupInterval = 3; // 3 minutes
+        private const int PurgeAfter = 30; // 30 minutes
+
+        // PREVIEW: unfortunately, I need dummy workspace since workspace services can be workspace specific
+        private static readonly Serializer s_serializer = new Serializer(new AdhocWorkspace(RoslynServices.HostServices, workspaceKind: "dummy").Services);
+
+        private readonly ConcurrentDictionary<int, AssetSource> _assetSources =
+            new ConcurrentDictionary<int, AssetSource>(concurrencyLevel: 4, capacity: 10);
+
+        private readonly ConcurrentDictionary<Checksum, Entry> _assets =
+            new ConcurrentDictionary<Checksum, Entry>(concurrencyLevel: 4, capacity: 10);
+
+        public AssetService()
+        {
+            Task.Factory.StartNew(CleanAssets, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default).Unwrap();
```

Comment:
Task.LongRunning shouldn't be necessary here.  When your task does run it will be for very short periods, and will relinquish control again almost immediately when it does the 'await delay'.  LongRunning is when you expect your task to need to occupy the use of the thread for significant periods of time.